lib/lwm2m_carrier/os/lwm2m_os.c

```diff
@@ -0,0 +1,235 @@
+	.sector_size = NVS_SECTOR_SIZE,
+	.sector_count = NVS_SECTOR_COUNT,
+	.offset = NVS_STORAGE_OFFSET,
+};
+
+int lwm2m_os_init(void)
+{
+	int err;
+
+	/* Initialize storage */
+	err = nvs_init(&fs, DT_FLASH_DEV_NAME);
+	if (err) {
+		return err;
+	}
+
+	return 0;
+}
+
+/* Memory management. */
+
+void *lwm2m_os_malloc(size_t size)
+{
+	return k_malloc(size);
+}
+
+void lwm2m_os_free(void *ptr)
+{
+	k_free(ptr);
+}
+
+/* Timing functions. */
+
+int64_t lwm2m_os_uptime_get(void)
+{
+	return k_uptime_get();
+}
+
+int lwm2m_os_sleep(int ms)
+{
+	return k_sleep(K_MSEC(ms));
+}
+
+/* OS functions */
+
+void lwm2m_os_sys_reset(void)
+{
+	spm_request_system_reboot();
+	CODE_UNREACHABLE;
+}
+
+uint32_t lwm2m_os_rand_get(void)
+{
+	return sys_rand32_get();
+}
+
+/* Non volatile storage */
+
+int lwm2m_os_storage_delete(uint16_t id)
+{
+	return nvs_delete(&fs, id);
+}
+
+int lwm2m_os_storage_read(uint16_t id, void *data, size_t len)
+{
+	return nvs_read(&fs, id, data, len);
+}
+
+int lwm2m_os_storage_write(uint16_t id, const void *data, size_t len)
+{
+	return nvs_write(&fs, id, data, len);
+}
+
+/* LWM2M timer, built on top of delayed works. */
+
+struct lwm2m_work {
+	struct k_delayed_work work_item;
+	lwm2m_os_timer_handler_t handler;
+};
+
+static struct lwm2m_work lwm2m_works[LWM2M_OS_MAX_TIMER_COUNT];
+
+static void work_handler(struct k_work *work)
+{
+	struct lwm2m_work *lwm2m_work =
+		CONTAINER_OF(work, struct lwm2m_work, work_item);
+
+	lwm2m_work->handler(lwm2m_work);
+}
+
+void *lwm2m_os_timer_get(lwm2m_os_timer_handler_t handler)
+{
+	struct lwm2m_work *work = NULL;
+
+	u32_t key = irq_lock();
+
+	/* Find free delayed work */
+	for (int i = 0; i < ARRAY_SIZE(lwm2m_works); i++) {
+		if (lwm2m_works[i].handler == NULL) {
+			work = &lwm2m_works[i];
+			work->handler = handler;
+			break;
+		}
+	}
+
+	irq_unlock(key);
+
+	if (work != NULL) {
+		k_delayed_work_init(&work->work_item, work_handler);
+	}
+
+	return work;
+}
+
+void lwm2m_os_timer_release(void *timer)
+{
+	struct lwm2m_work *work = (struct lwm2m_work *)timer;
+
+	if (!PART_OF_ARRAY(lwm2m_works, work)) {
+		return;
+	}
+
+	work->handler = NULL;
+}
+
+int lwm2m_os_timer_start(void *timer, int32_t timeout)
+{
+	struct lwm2m_work *work = (struct lwm2m_work *)timer;
+
+	if (!PART_OF_ARRAY(lwm2m_works, work)) {
+		return -EINVAL;
+	}
+
+	/* Zephyr's timing subsystem uses positive integers so the
+	 * largest tick count that can be represented is 31 bit large.
+	 * Avoid requesting timeouts larger than that, or they will expire
+	 * immediately.
+	 *
+	 * max_timeout_sec = int max - 1 / ticks per sec
+	 */
+	const int max_timeout_sec =
+		(INT32_MAX - 1) / CONFIG_SYS_CLOCK_TICKS_PER_SEC;
+
+	/* `timeout` is expressed in milliseconds,
+	 * use K_SECONDS to specify the correct unit.
+	 */
+	timeout = MIN(K_SECONDS(max_timeout_sec), timeout);
+
+	/* FIXME: return an error if the timeout requested is too large. */
+	return k_delayed_work_submit(&work->work_item, K_MSEC(timeout));
```

Comment:
I don't get it.

Both the wrapper, and the wrappee are expressing timeouts in signed 32 bit milliseconds, so why do we need to do any correction?

Reply:
Internally the timeout is converted in number of ticks and those are signed as far I understand.

Reply:
> Internally the timeout is converted in number of ticks

What is working with ticks? **k_delayed_work_submit** definitely works with milliseconds, and AFAICT **lwm2m_os_timer_start** does as well, or else you would get the wrong behaviour when timeout == 1 for instance.

Reply:
My suggestion is to delete line 171 and line 165.

Reply:
See [here](https://github.com/zephyrproject-rtos/zephyr/blob/82c25587a2bc7610530154df4d984bdfa2add594/kernel/work_q.c#L110). Notice:
- `void z_add_timeout(struct _timeout *to, _timeout_func_t fn, s32_t ticks)` and
- ` s32_t z_ms_to_ticks(s32_t ms)` signatures.

In particular tick counting is signed.

Reply:
Right, k_delayed_work_submit uses ticks internally, but externally both k_delayed_work_submit and lwm2m_os_timer_start uses ms, so we don't need to touch the timeout value in this function.

Reply:
The code here detects if any timeout in milliseconds can be expressed correctly in ticks before submitting the work item, it's not doing a conversion. When a timeout of 1 is requested, it will timeout in one millisecond.
Note that: `#define K_MSEC(ms)     (ms)`,
I have just added that to make the unit very explicit.

Reply:
Isn't it the responsiblity of **k_delayed_work_submit** to detect this? How is this a unique requirement of **lwm2m_os_timer_start**?

What happens when it can't be expressed and why can't it always be expressed?

Reply:
This is a limitation of Zephyr's current timing system, as far as I understand..
If a timeout that becomes negative when translated into ticks (`int`) is treated as if it had expired and the delayed work will be fired off at the earliest.

For example: `K_HOURS(20)` , would be translated to ticks as:
`20 * 60 * 60 * CONFIG_SYS_CLOCK_TICKS_PER_SEC (32768) = 0x8CA00000`.
That number is negative and submitting a work item with a delay of `K_HOURS(20)` will make the work item expire immediately instead of in twenty hours.

Reply:
This should be reported as a bug and a reference to the bugreport should be listed next to our workaround code.

Reply:
Perhaps I can could a reference to this: https://github.com/zephyrproject-rtos/zephyr/issues/9663? Otherwise I could create a new issue and reference that, but I think there are already many issues about this topic.

Reply:
That is not relevant. A new issue should be created, it will be closed as duplicate if appropriate.

Reply:
Done.